### PR TITLE
Support retainkeys strategy for new merge code

### PR DESCRIPTION
**Pull request by pwittrock**
- Some prefactoring for retainkeys
- Add retainkeys strategy

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/apply/parse/openapi.go
Comment (apelisse) [inline pkg/kubectl/apply/parse/openapi.go]: I don't understand what this block of code is doing?

File: pkg/kubectl/apply/parse/util.go
Comment (apelisse) [inline pkg/kubectl/apply/parse/util.go]: Would it be simpler to remove that `if` statement, since `m.MergeType` is always equal to `strings.Split(strategy, ",")`, even if there is no comma?

File: pkg/kubectl/apply/strategy/strategic_visitor.go
Comment (apelisse) [inline pkg/kubectl/apply/strategy/strategic_visitor.go]: Adding a new strategy implies that you have 3 places to update here (type, map, list). Shouldn't this be a dispatch rather than a "if/switch" statement?

EDIT: I understand that it might not be that easy ...

File: pkg/kubectl/apply/strategy/retain_keys_visitor.go
Comment (apelisse) [inline pkg/kubectl/apply/strategy/retain_keys_visitor.go]: This is really wonderful. A new strategy in 77 lines, copyright and imports included.

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (apelisse) [inline pkg/kubectl/apply/strategy/retain_keys_test.go]: Can this happen, and is that what we want? A controller has added a new volume for some reason. The user can come-up with the same name and just override it?

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (apelisse) [inline pkg/kubectl/apply/strategy/retain_keys_test.go]: Can we have both `hostPath` and `emptyDir`?

**Review by apelisse — COMMENTED** [review]
Great!

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (pwittrock) [reply pkg/kubectl/apply/strategy/retain_keys_test.go]: Yes.  Or someone changed the Volume with edit for example.

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (pwittrock) [reply pkg/kubectl/apply/strategy/retain_keys_test.go]: No, this will fail validation on the server side, but the merge logic is correct here.

Comment (apelisse) [reply pkg/kubectl/apply/strategy/retain_keys_test.go]: Cool, that's what I expected, just wanted to be sure.

File: pkg/kubectl/apply/parse/openapi.go
Comment (pwittrock) [reply pkg/kubectl/apply/parse/openapi.go]: added comments

File: pkg/kubectl/apply/parse/util.go
Comment (pwittrock) [reply pkg/kubectl/apply/parse/util.go]: done

**Comment by pwittrock** [comment]
PTAL

**Review by apelisse — APPROVED** [review]
I get it now :-)

**Comment by apelisse** [comment]
LGTM for me, @mengqiy please `/lgtm` when you're done.

File: pkg/kubectl/apply/parse/openapi.go
Comment (mengqiy) [inline pkg/kubectl/apply/parse/openapi.go]: Have we enforce `retainKeys` always comes before `merge`?
Old code has some logic to handle this: https://github.com/kubernetes/kubernetes/blob/afae7bccf56f85d9faf5ca4f6b702af53770f0e1/staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go#L2102-L2110

File: pkg/kubectl/apply/parse/util.go
Comment (mengqiy) [inline pkg/kubectl/apply/parse/util.go]: Some question here.

File: pkg/kubectl/apply/strategy/retain_keys_visitor.go
Comment (mengqiy) [inline pkg/kubectl/apply/strategy/retain_keys_visitor.go]: > MergeMap returns and error.

Nit: something wrong with the comment.

File: pkg/kubectl/apply/strategy/retain_keys_visitor.go
Comment (mengqiy) [inline pkg/kubectl/apply/strategy/retain_keys_visitor.go]: > MergeMap returns and error.

Nit: something wrong with the comment.

File: pkg/kubectl/apply/strategy/retain_keys_visitor.go
Comment (mengqiy) [inline pkg/kubectl/apply/strategy/retain_keys_visitor.go]: > MergeMap returns and error.

Nit: something wrong with the comment.

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (mengqiy) [inline pkg/kubectl/apply/strategy/retain_keys_test.go]: Should this be `strategy: {}`?

I think `{}` means it's an empty map, but unspecified means it's null.

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (mengqiy) [inline pkg/kubectl/apply/strategy/retain_keys_test.go]: Same here.

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (mengqiy) [inline pkg/kubectl/apply/strategy/retain_keys_test.go]: Same here.

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (mengqiy) [inline pkg/kubectl/apply/strategy/retain_keys_test.go]: Same here.

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (mengqiy) [inline pkg/kubectl/apply/strategy/retain_keys_test.go]: Same here.

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (mengqiy) [inline pkg/kubectl/apply/strategy/retain_keys_test.go]: Same here.

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (mengqiy) [reply pkg/kubectl/apply/strategy/retain_keys_test.go]: BTW how do we handle `null` in the new code?

**Review by mengqiy — COMMENTED** [review]
Some minor comments

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (pwittrock) [reply pkg/kubectl/apply/strategy/retain_keys_test.go]: null uses the VisitEmpty method.

File: pkg/kubectl/apply/parse/openapi.go
Comment (pwittrock) [reply pkg/kubectl/apply/parse/openapi.go]: I would rather keep the code generic instead of encoding specific business logic into it.  I could imagine a `merge,replace` strategy being reasonable as well.

File: pkg/kubectl/apply/parse/util.go
Comment (pwittrock) [reply pkg/kubectl/apply/parse/util.go]: See above

File: pkg/kubectl/apply/strategy/retain_keys_test.go
Comment (pwittrock) [reply pkg/kubectl/apply/strategy/retain_keys_test.go]: I think these are fine for the test.  both null or empty map execute the logic.

**Comment by mengqiy** [comment]
LGTM

Feel free to re-apply the lgtm label after fixing the nit in comments.

**Pull request merged**
